packages/packagers/css/src/CSSPackager.js

```diff
@@ -106,77 +134,112 @@ export default (new Packager({
                   }
                 }
 
-                if (media.length) {
-                  return `@media ${media.join(', ')} {\n${css}\n}\n`;
-                }
-
                 return css;
               }),
-              bundle.env.sourceMap && asset.getMapBuffer(),
+              bundle.env.sourceMap ? asset.getMap() : null,
             ]);
           }
         });
       },
     });
 
-    let outputs = await queue.run();
-    let contents = '';
+    let outputs = new Map(
+      (await queue.run()).map(([asset, code, map]) => [asset, [code, map]]),
+    );
     let map = new SourceMap(options.projectRoot);
-    let lineOffset = 0;
 
-    for (let url of hoistedImports) {
-      contents += `@import "${url}";\n`;
-      lineOffset++;
-    }
+    let res = await bundleAsync({
+      filename: nullthrows(entry),
+      sourceMap: !!bundle.env.sourceMap,
+      resolver: {
+        resolve(specifier) {
+          return specifier;
+        },
+        async read(file) {
+          if (file === bundle.id) {
+            return entryContents;
+          }
 
-    for (let [asset, code, mapBuffer] of outputs) {
-      contents += code + '\n';
-      if (bundle.env.sourceMap) {
-        if (mapBuffer) {
-          map.addBuffer(mapBuffer, lineOffset);
-        } else {
-          map.addEmptyMap(
-            path
-              .relative(options.projectRoot, asset.filePath)
-              .replace(/\\+/g, '/'),
-            code,
-            lineOffset,
-          );
-        }
+          let asset = assetsByPlaceholder.get(file);
+          if (!asset) {
+            return '';
+          }
+          let [code, map] = nullthrows(outputs.get(asset));
+          if (map) {
+            let sm = await map.stringify({format: 'inline'});
+            invariant(typeof sm === 'string');
+            code += `\n/*# sourceMappingURL=${sm} */`;
+          }
+          return code;
```

Comment:
Isn't this repeated stringification, inlining and parsing of the sourcemap pretty slow?

Reply:
probably not ideal. once we move to rust we can just pass around a parcel source map instance. lightningcss uses the same library internally but the JS API still returns it as a JSON buffer.